src/scripts/getRefreshToken.ts

```diff
@@ -1,43 +1,45 @@
 import fetch from "node-fetch";
-import { DesktopIAPOptions, Fetcher } from "../types";
+import { DesktopIAPOptions } from "../types";
+import url from "url";
 
 export function openBrowser(browser: (url: string) => void, url: string): void {
   browser(url);
 }
 
 export async function getRefreshToken(
-  browser: (url: string) => void,
-  inputHandler: (question: string) => Promise<string>,
-  fetcher: Fetcher = fetch,
+  redirectUrl: string,
+  fetcher: Function = fetch,
   options: DesktopIAPOptions,
+  redirectUri: string,
 ): Promise<string> {
-  const url = `https://accounts.google.com/o/oauth2/v2/auth?client_id=${options.clientId}&response_type=code&scope=openid%20email&access_type=offline&redirect_uri=urn:ietf:wg:oauth:2.0:oob`;
-
-  console.log("Open this URL if your browser didn't open automatically\n", url);
-
-  openBrowser(browser, url);
-
-  // await input from user
-  const loginToken = await inputHandler(
-    "Please input the token you received in the browser",
-  );
-
-  const oauthTokenBaseUrl = "https://www.googleapis.com/oauth2/v4/token";
-
-  const body = {
-    code: loginToken,
-    client_id: options.clientId,
-    client_secret: options.clientSecret,
-    redirect_uri: "urn:ietf:wg:oauth:2.0:oob",
-    grant_type: "authorization_code",
-  };
-
-  const request = await fetcher(oauthTokenBaseUrl, {
-    body: JSON.stringify(body),
-    method: "POST",
-  });
-
-  const response = await request.json();
-
-  return String(response["refresh_token"]);
+  let queryParam = url.parse(redirectUrl, true).query;
+  const loginToken = queryParam.code;
```

Comment:
Could there be some parsing/validation of user input?

Reply:
What do you mean by validation of "user input"? Should I check if redirect_uri contains code as queryParam?
I create this MR by using google example -> https://github.com/googleapis/google-api-nodejs-client/blob/main/samples/oauth2.js

Reply:
I understand you are using the example provided by Google, but query parameter without any parsing or validation can be "dangerous".

For example, is the code supposed to have a specific length, match some regular expression, belong to a given dictionary of possible cases?

https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html

Reply:
I try to Google it, but I do not find anything about Google OAuth. OAuth specification do not say anything particular about authorisation code. So it could be any length and contains any characters (it will contains only 'url safe'). I could probably make check if it is ASCII character. Code looks like this:
4/WwBf2eVqGtlxmBbPMGymGi18OJCj3JFMhXouWO1jrdnyWGqH4pcNbqIfxWQnTWA_AG-sX3nRqEUn_Z2x0WGHhQg